### PR TITLE
Correct battle turn order in rpg2k battle - previously Albeleon 2.42

### PR DESCRIPTION
**Pull request by fmatthew5876**
From Cherry:
> RPG_RT defines something like a "battle agility" for each turn which is AGI+X. X is a random number below (AGI/4 + 3). However, if the attack is a basic attack with a weapon that has the "Start with Initiative" checkbox, this puts it into a class of its own which follows the same rules but can rival only other attacks with such weapons (I think this is done by simply adding a huge number like 100000 to the value in such a case). The turn order is then decided by this "battle agility", high to low.

### CODE REVIEW

**Comment by fmatthew5876** [comment]
I've tested this and confirmed the behavior matches RPG_RT

**Comment by fmatthew5876** [comment]
jenkins: test this please

**Comment by Ghabry** [comment]
This technique for "::Type" propagation through the base constructor is pretty neat btw.

File: src/scene_battle_rpg2k.cpp [hunk 1]
```diff
@@ -849,24 +849,19 @@ void Scene_Battle_Rpg2k::SelectPreviousActor() {
...
-	return first->GetAgi() > second->GetAgi();
-}
-
 void Scene_Battle_Rpg2k::CreateExecutionOrder() {
-	std::sort(battle_actions.begin(), battle_actions.end(), BattlerSort);
+	// Define random Agility. Must be done outside of the sort function because of the "strict weak ordering" property, so the sort is consistent
+	for (auto battler : battle_actions) {
+		int battle_order = battler->GetAgi() + Utils::GetRandomNumber(0, battler->GetAgi() / 4 + 3);
+		if (battler->GetBattleAlgorithm()->GetType() == Game_BattleAlgorithm::Type::Normal && battler->HasPreemptiveAttack()) {
+			battle_order += 100000;
```
Comment (Ghabry) [inline src/scene_battle_rpg2k.cpp]: Could you add a comment that this is an "arbitrary huge number"?

**Comment by fmatthew5876** [comment]
> 
> 
> This technique for "::Type" propagation through the base constructor is pretty neat btw.

This commit is from battle 5. I like this approach because you can make `GetType()` non-virtual and optimize a lot of usage.

I've added the requested comment.

File: src/scene_battle_rpg2k.cpp [hunk 2]
```diff
@@ -849,24 +849,20 @@ void Scene_Battle_Rpg2k::SelectPreviousActor() {
...
-	return first->GetAgi() > second->GetAgi();
-}
-
 void Scene_Battle_Rpg2k::CreateExecutionOrder() {
-	std::sort(battle_actions.begin(), battle_actions.end(), BattlerSort);
+	// Define random Agility. Must be done outside of the sort function because of the "strict weak ordering" property, so the sort is consistent
+	for (auto battler : battle_actions) {
+		int battle_order = battler->GetAgi() + Utils::GetRandomNumber(0, battler->GetAgi() / 4 + 3);
+		if (battler->GetBattleAlgorithm()->GetType() == Game_BattleAlgorithm::Type::Normal && battler->HasPreemptiveAttack()) {
+			// This is an arbitrarily large number to separate preemptive vs non-premptive battlers
```
Comment (carstene1ns) [inline src/scene_battle_rpg2k.cpp]: typo. `non-preemptive`

Comment (fmatthew5876) [reply src/scene_battle_rpg2k.cpp]: Fixed

**Pull request merged**